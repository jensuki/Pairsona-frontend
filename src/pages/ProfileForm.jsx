import { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { usePlacesWidget } from 'react-google-autocomplete';
import getProfilePic from '../helpers/getProfilePic';
import Loading from '../components/Loading';
import useAuth from '../hooks/useAuth';
import Api from '../helpers/api';
import '../styles/Forms.css';

/** component to update current users profile */
const ProfileForm = () => {
    const { currUser, setCurrUser } = useAuth();

    const navigate = useNavigate();

    // state
    const [formData, setFormData] = useState({
        username: '',
        firstName: '',
        lastName: '',
        email: '',
        birthDate: '',
        location: '',
        bio: '',
        profilePic: null,
        password: ''
    });
    const [picPreview, setPicPreview] = useState(getProfilePic(null));
    const [error, setError] = useState(null);
    const [loading, setLoading] = useState(false);

    // usePlacesWidget reference from 'https://www.npmjs.com/package/react-google-autocomplete'
    // destructure ref from useplaceswidget hook to get input ref
    const { ref: locationInputRef } = usePlacesWidget({
        apiKey: import.meta.env.VITE_GOOGLE_PLACES_API_KEY,
        onPlaceSelected: (place) => {
            // update formdata if valid location is selected
            if (place && place.formatted_address) {
                setFormData((fData) => ({
                    ...fData,
                    location: place.formatted_address // set new location
                }))
                setError(null);
            }
        },
        options: {
            types: ['(cities)'],
            componentRestrictions: { country: 'us' }
        } // restrict to cities in us
    })

    // populate formdata whenever curr user changes
    useEffect(() => {
        if (currUser) {
            setFormData({
                username: currUser.username || '',
                firstName: currUser.firstName || '',
                lastName: currUser.lastName || '',
                email: currUser.email || '',
                birthDate: currUser.birthDate || '',
                location: currUser.location || '',
                bio: currUser.bio || '',
                profilePic: null,
                password: ''
            });
            setPicPreview(getProfilePic(currUser.profilePic));
        }
    }, [currUser]);

    // handle form input changes
    const handleChange = (e) => {
        const { name, value } = e.target;
        setFormData((fData) => ({
            ...fData,
            [name]: value,
        }));
        setError(null);
    };

    // handle new profile picture upload
    const handleFileChange = (e) => {
        const file = e.target.files[0];
        if (file) {
            setFormData((fData) => ({
                ...fData,
                profilePic: file,
            }));
            setPicPreview(URL.createObjectURL(file));
        }
    };

    // handle resetting pic to default
    const handleResetPic = () => {
        setFormData((fData) => ({
            ...fData,
            profilePic: null, // clear profile pic
            resetProfilePic: 'true'// set true for backend to remove image
        }));
        setPicPreview(getProfilePic(null));
    }

    // handle form submission
    const handleSubmit = async (e) => {
        e.preventDefault();
        setError(null);
        setLoading(true);

        try {
            const formDataToSend = new FormData();

            // append all valid fields
            for (const [key, value] of Object.entries(formData)) {
                if (value !== undefined && value !== null) {
                    formDataToSend.append(key, value);
                }
            }

            // send the data to the backend
            const updatedUser = await Api.updateUser(formData.username, formDataToSend);

            // update the current users state
            setCurrUser((prevUser) => ({
                ...prevUser,
                ...updatedUser,
            }));

            navigate('/profile');
        } catch (err) {
            setError(err.message);

            setLoading(false);
        }
    };

    if (loading) return <Loading />;

    return (
        <div className="FormContainer ProfileForm">
            <h1>Update Your Profile</h1>
            <form onSubmit={handleSubmit} encType="multipart/form-data">

                {/* display error if exists */}
                {error && <p className="form-error">{error}</p>}

                {/* profile pic preview */}
                <div className="form-group">
                    <label>Current Profile Picture</label>
                    <div className="pic-preview">
                        <img
                            src={picPreview}
                            alt="Profile Pic Preview"
                            width="150"
                            height="150"
                        />
                    </div>
                </div>

                {/* change profile pic input */}
                <div className="form-group">
                    <label htmlFor="profilePic">Change Profile Picture</label>
                    <input
                        type="file"
                        name="profilePic"
                        id="profilePic"
                        onChange={handleFileChange}
                    />
                </div>

                {/* reset profile pic */}
                <div className="form-group">
                    <button
                        type="button"
                        className="reset-pic"
                        onClick={handleResetPic}>
                        Reset Profile Picture
                    </button>
                </div>

                {/* username (read only) */}
                <div className="form-group">
                    <label htmlFor="username">Username</label>
                    <input
                        type="text"
                        name="username"
                        id="username"
                        defaultValue={formData.username}
                        autoComplete="username"
                        disabled
                    />
                </div>

                {/* first name */}
                <div className="form-group">
                    <label htmlFor="firstName">First Name</label>
                    <input
                        type="text"
                        name="firstName"
                        id="firstName"
                        value={formData.firstName}
                        onChange={handleChange}
                        required
                    />
                </div>

                {/* last name */}
                <div className="form-group">
                    <label htmlFor="lastName">Last Name</label>
                    <input
                        type="text"
                        name="lastName"
                        id="lastName"
                        value={formData.lastName}
                        onChange={handleChange}
                        required
                    />
                </div>

                {/* email */}
                <div className="form-group">
                    <label htmlFor="email">Email</label>
                    <input
                        type="email"
                        name="email"
                        id="email"
                        value={formData.email}
                        onChange={handleChange}
                        required
                    />
                </div>

                {/* birth date */}
                <div className="form-group">
                    <label htmlFor="birthDate">Birth Date</label>
                    <input
                        type="date"
                        name="birthDate"
                        id="birthDate"
                        value={formData.birthDate}
                        disabled
                    />
                </div>

                {/* location input with autocomplete */}
                <div className="form-group">
                    <label htmlFor="location">Location</label>
                    <input
                        ref={locationInputRef} // use input ref
                        type="text"
                        id="location"
                        placeholder="Type a city"
                        defaultValue={formData.location}
                        required
                    />
                </div>

                {/* bio */}
                <div className="form-group">
                    <label htmlFor="bio">Bio</label>
                    <textarea
                        name="bio"
                        id="bio"
                        value={formData.bio}
                        onChange={handleChange}
                        placeholder="Tell us about yourself"
                    />
                </div>

                {/* password */}
                <div className="form-group">
                    <label htmlFor="password">Password</label>
                    <input
                        type="password"
                        name="password"
                        id="password"
                        value={formData.password}
                        onChange={handleChange}
                        placeholder="Enter password to confirm changes"
                        autoComplete="current-password"
                        required
                    />
                </div>

                <div className="form-actions">
                    <button type="submit" className="profile-btn">Save Changes</button>
                    <button
                        type="button"
                        className="cancel-btn"
                        onClick={() => navigate('/profile')}
                    >
                        Cancel
                    </button>
                </div>
            </form>
        </div>
    );
};

export default ProfileForm;
