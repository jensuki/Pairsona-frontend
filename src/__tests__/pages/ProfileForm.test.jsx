import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../test_common/setupTests';
import ProfileForm from '../../pages/ProfileForm';
import Api from '../../helpers/api';
import { describe, test, expect, vi, beforeEach } from 'vitest';
import { demoUser } from '../../test_common/mockData';

// mock useAuth hook
vi.mock('../../hooks/useAuth', () => ({
    default: vi.fn(() => ({
        currUser: demoUser, // set demouser as curr user
    }))
}));

// mock api func for update
vi.mock('../../helpers/api', () => ({
    default: {
        updateUser: vi.fn(),
    },
}));

// mock profile pic logic
vi.mock('../helpers/getProfilePic', () => ({
    default: vi.fn((pic) => pic || '/assets/default_pic.png'),
}));

// helper
const renderProfileForm = () => {
    return render(
        <MemoryRouter>
            <UserProvider>
                <ProfileForm />
            </UserProvider>
        </MemoryRouter>
    );
};

describe('ProfileForm component', () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    test('renders without crashing and populates initial values', async () => {
        renderProfileForm();

        // check that the initial values are displayed
        expect(screen.getByLabelText('First Name')).toHaveValue(demoUser.firstName);
        expect(screen.getByLabelText('Last Name')).toHaveValue(demoUser.lastName);
        expect(screen.getByLabelText('Email')).toHaveValue(demoUser.email);
    });

    test('matches snapshot', () => {
        const { asFragment } = renderProfileForm();
        expect(asFragment()).toMatchSnapshot();
    })

    test('updates input fields correctly', () => {
        renderProfileForm();

        // update first and last name
        const firstnameInput = screen.getByLabelText('First Name');
        const lastNameInput = screen.getByLabelText('Last Name');

        fireEvent.change(firstnameInput, { target: { value: 'Updated First Name' } });
        fireEvent.change(lastNameInput, { target: { value: 'Updated Last Name' } });

        expect(firstnameInput.value).toBe('Updated First Name');
        expect(lastNameInput.value).toBe('Updated Last Name');
    })

    test('handles profile pic upload', async () => {
        renderProfileForm();

        // upload a new img file
        const file = new File(['profile-pic'], 'profile-pic.png', { type: 'image/png' });
        const fileInput = screen.getByLabelText('Change Profile Picture');

        fireEvent.change(fileInput, { target: { files: [file] } });

        // wait for the preview to load
        await waitFor(() => {
            expect(screen.getByAltText('Profile Pic Preview')).toBeInTheDocument();
        })
    })

    test('matches snapshot after updating profile pic', async () => {
        const { asFragment } = renderProfileForm();

        const file = new File(['profile-pic'], 'new-pic.png', { type: 'image/png' });
        const fileInput = screen.getByLabelText('Change Profile Picture');

        fireEvent.change(fileInput, { target: { files: [file] } });

        await waitFor(() => {
            expect(asFragment()).toMatchSnapshot();
        })
    })

    test('displays error message on failed form submission', async () => {
        // mock error msg
        Api.updateUser.mockRejectedValueOnce(new Error('Update Failed'));

        renderProfileForm();

        // fill out required password field
        const passwordInput = screen.getByLabelText('Password');
        fireEvent.change(passwordInput, { target: { value: 'password123' } });

        // click save changes
        const saveBtn = screen.getByText('Save Changes');
        fireEvent.click(saveBtn);

        // wait for error msg to appear
        await waitFor(() => {
            expect(screen.getByText('Update Failed')).toBeInTheDocument();
        })
    })


    test('handles successful form submission', async () => {
        Api.updateUser.mockResolvedValueOnce({
            ...demoUser,
            firstName: 'Updated Name'
        })

        renderProfileForm();

        const firstnameInput = screen.getByLabelText('First Name');
        const passwordInput = screen.getByLabelText('Password');

        // update fields
        fireEvent.change(firstnameInput, { target: { value: 'Updated Name' } })
        fireEvent.change(passwordInput, { target: { value: 'password123' } });

        // submit form
        const saveBtn = screen.getByText('Save Changes');
        fireEvent.click(saveBtn);

        await waitFor(() => {
            expect(Api.updateUser).toHaveBeenCalledWith(
                demoUser.username,
                expect.any(FormData)
            )
        })
    })
});