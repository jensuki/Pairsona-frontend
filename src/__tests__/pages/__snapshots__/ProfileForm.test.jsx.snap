// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProfileForm component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="FormContainer ProfileForm"
  >
    <h1>
      Update Your Profile
    </h1>
    <form
      enctype="multipart/form-data"
    >
      <div
        class="form-group"
      >
        <label>
          Current Profile Picture
        </label>
        <div
          class="pic-preview"
        >
          <img
            alt="Profile Pic Preview"
            height="150"
            src="/assets/default_pic.png"
            width="150"
          />
        </div>
      </div>
      <div
        class="form-group"
      >
        <label
          for="profilePic"
        >
          Change Profile Picture
        </label>
        <input
          id="profilePic"
          name="profilePic"
          type="file"
        />
      </div>
      <div
        class="form-group"
      >
        <button
          class="reset-pic"
          type="button"
        >
          Reset Profile Picture
        </button>
      </div>
      <div
        class="form-group"
      >
        <label
          for="username"
        >
          Username
        </label>
        <input
          autocomplete="username"
          disabled=""
          id="username"
          name="username"
          type="text"
          value="testuser"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="firstName"
        >
          First Name
        </label>
        <input
          id="firstName"
          name="firstName"
          required=""
          type="text"
          value="Test"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="lastName"
        >
          Last Name
        </label>
        <input
          id="lastName"
          name="lastName"
          required=""
          type="text"
          value="User"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="email"
        >
          Email
        </label>
        <input
          id="email"
          name="email"
          required=""
          type="email"
          value="testuser@example.com"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="birthDate"
        >
          Birth Date
        </label>
        <input
          disabled=""
          id="birthDate"
          name="birthDate"
          type="date"
          value="2000-01-01"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="location"
        >
          Location
        </label>
        <input
          id="location"
          placeholder="Type a city"
          required=""
          type="text"
          value="New York, NY"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="bio"
        >
          Bio
        </label>
        <textarea
          id="bio"
          name="bio"
          placeholder="Tell us about yourself"
        >
          Loves personality tests
        </textarea>
      </div>
      <div
        class="form-group"
      >
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="current-password"
          id="password"
          name="password"
          placeholder="Enter password to confirm changes"
          required=""
          type="password"
          value=""
        />
      </div>
      <div
        class="form-actions"
      >
        <button
          class="profile-btn"
          type="submit"
        >
          Save Changes
        </button>
        <button
          class="cancel-btn"
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`ProfileForm component > matches snapshot after updating profile pic 1`] = `
<DocumentFragment>
  <div
    class="FormContainer ProfileForm"
  >
    <h1>
      Update Your Profile
    </h1>
    <form
      enctype="multipart/form-data"
    >
      <div
        class="form-group"
      >
        <label>
          Current Profile Picture
        </label>
        <div
          class="pic-preview"
        >
          <img
            alt="Profile Pic Preview"
            height="150"
            src="mock-preview-url"
            width="150"
          />
        </div>
      </div>
      <div
        class="form-group"
      >
        <label
          for="profilePic"
        >
          Change Profile Picture
        </label>
        <input
          id="profilePic"
          name="profilePic"
          type="file"
        />
      </div>
      <div
        class="form-group"
      >
        <button
          class="reset-pic"
          type="button"
        >
          Reset Profile Picture
        </button>
      </div>
      <div
        class="form-group"
      >
        <label
          for="username"
        >
          Username
        </label>
        <input
          autocomplete="username"
          disabled=""
          id="username"
          name="username"
          type="text"
          value="testuser"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="firstName"
        >
          First Name
        </label>
        <input
          id="firstName"
          name="firstName"
          required=""
          type="text"
          value="Test"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="lastName"
        >
          Last Name
        </label>
        <input
          id="lastName"
          name="lastName"
          required=""
          type="text"
          value="User"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="email"
        >
          Email
        </label>
        <input
          id="email"
          name="email"
          required=""
          type="email"
          value="testuser@example.com"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="birthDate"
        >
          Birth Date
        </label>
        <input
          disabled=""
          id="birthDate"
          name="birthDate"
          type="date"
          value="2000-01-01"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="location"
        >
          Location
        </label>
        <input
          id="location"
          placeholder="Type a city"
          required=""
          type="text"
          value="New York, NY"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          for="bio"
        >
          Bio
        </label>
        <textarea
          id="bio"
          name="bio"
          placeholder="Tell us about yourself"
        >
          Loves personality tests
        </textarea>
      </div>
      <div
        class="form-group"
      >
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="current-password"
          id="password"
          name="password"
          placeholder="Enter password to confirm changes"
          required=""
          type="password"
          value=""
        />
      </div>
      <div
        class="form-actions"
      >
        <button
          class="profile-btn"
          type="submit"
        >
          Save Changes
        </button>
        <button
          class="cancel-btn"
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
